### PR TITLE
Personalised FedReID

### PR DESCRIPTION
**Pull request by codergan**
support personalised clustering training for FedReID

### CODE REVIEW

File: client.py [hunk 1]
```diff
@@ -60,12 +64,12 @@ def train(self, federated_model, use_cuda):
...
-                    inputs = Variable(inputs.cuda().detach())
-                    labels = Variable(labels.cuda().detach())
-                else:
-                    inputs, labels = Variable(inputs), Variable(labels)
-                
+                # if use_cuda:
+                #     inputs = Variable(inputs.cuda().detach())
+                #     labels = Variable(labels.cuda().detach())
+                # else:
+                #     inputs, labels = Variable(inputs), Variable(labels)
```
Comment (weimingwill) [inline client.py:71]: Can just remove these comments

File: client.py [hunk 2]
```diff
@@ -21,25 +21,30 @@ def __init__(self, cid, data, device, project_dir, model_name, local_epoch, lr,
...
+        else:
+            self.model.load_state_dict(federated_model.state_dict())
         self.model.classifier.classifier = self.classifier
         self.old_classifier = copy.deepcopy(self.classifier)
         self.model = self.model.to(self.device)
-
+        self.model.train(True)
         optimizer = get_optimizer(self.model, self.lr)
-        scheduler = lr_scheduler.StepLR(optimizer, step_size=40, gamma=0.1)
+        # scheduler = lr_scheduler.StepLR(optimizer, step_size=40, gamma=0.1)
```
Comment (weimingwill) [inline client.py:47]: Why we don't need the scheduler anymore?

File: evaluate.py [hunk 3]
```diff
@@ -2,11 +2,11 @@
...
-import argparse
-parser = argparse.ArgumentParser(description='Training')
-parser.add_argument('--result_dir', default='.', type=str)
-parser.add_argument('--dataset', default='no_dataset', type=str)
-args = parser.parse_args()
+# import argparse
+# parser = argparse.ArgumentParser(description='Training')
+# parser.add_argument('--result_dir', default='.', type=str)
+# parser.add_argument('--dataset', default='no_dataset', type=str)
+# args = parser.parse_args()
```
Comment (weimingwill) [inline evaluate.py:9]: Remove these comments

File: evaluate.py [hunk 4]
```diff
@@ -60,32 +60,39 @@ def compute_mAP(index, good_index, junk_index):
...
-gallery_feature = gallery_feature.cuda()
+    query_feature = torch.FloatTensor(result['query_f'])
+    query_cam = np.array(result['query_cam'])
+    query_label = np.array(result['query_label'])
+    gallery_feature = torch.FloatTensor(result['gallery_f'])
+    gallery_cam = np.array(result['gallery_cam'])
+    gallery_label = np.array(result['gallery_label'])
+    # print(type(query_feature),query_feature[:3])
+    # print(type(query_cam),query_cam[:3])
+    # print(type(query_label),query_label[:3])
```
Comment (weimingwill) [inline evaluate.py:78]: Remove these debugging prints

File: evaluate.py [hunk 5]
```diff
@@ -60,32 +60,39 @@ def compute_mAP(index, good_index, junk_index):
...
-query_cam = result['query_cam'][0]
-query_label = result['query_label'][0]
-gallery_feature = torch.FloatTensor(result['gallery_f'])
-gallery_cam = result['gallery_cam'][0]
-gallery_label = result['gallery_label'][0]
+def testing_model(result, dataset):
+    # result = scipy.io.loadmat(file_path)
+    # print("========= after loading ==========")
+    # for i in result:
+    #     print(i, np.array(result[i]).shape)
```
Comment (weimingwill) [inline evaluate.py:68]: Remove these comments

File: main.py [hunk 6]
```diff
@@ -53,24 +54,87 @@
...
+    else:
+        cdw = "Ncdw"
+    if args.kd:
+        kd = "kd"
+    else:
+        kd = "Nkd"
+    if args.regularization:
+        reg = "reg"
+    else:
+        reg = "Nreg"
```
Comment (weimingwill) [inline main.py:110]: Are these `if else` from line 95 - 110 only for setting the name for the checkpoint? It would be better to put them into a function.

File: main.py [hunk 7]
```diff
@@ -53,24 +54,87 @@
 parser.add_argument('--multiple_scale',default='1', type=str,help='multiple_scale: e.g. 1 1,1.1  1,1.1,1.2')
 parser.add_argument('--test_dir',default='all',type=str, help='./test_data')
 
+parser.add_argument('--resume_epoch', default=0, type=int, help='resume from which epoch, if 0, no resume')
```
Comment (weimingwill) [inline main.py:57]: Shouldn't it be resume round?

File: running.sh [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export PYTHONPATH=$PYTHONPATH:$pwd
```
Comment (weimingwill) [inline running.sh:1]: Put all these scripts for running into a separate folder `script`

File: server.py [hunk 9]
```diff
@@ -1,16 +1,20 @@
...
 import torch.nn as nn
 import torch
 import scipy.io
 import copy
+import numpy as np
 from data_utils import ImageDataset
-import random
 import torch.optim as optim
 from torchvision import datasets
+# from finch import FINCH
```
Comment (weimingwill) [inline server.py:14]: remove comment

File: server.py [hunk 10]
```diff
@@ -68,45 +74,97 @@ def __init__(self, clients, data, device, project_dir, model_name, num_of_client
...
+
+
         if epoch==0:
             self.L0 = torch.Tensor(loss) 
 
         avg_loss = sum(loss) / self.num_of_clients
 
         print("==============================")
-        print("number of clients used:", len(models))
+        # print("number of clients used:", len(models))
```
Comment (weimingwill) [inline server.py:138]: Remove comment

**Review by weimingwill — CHANGES_REQUESTED** [review]
